### PR TITLE
feat(discount-codes): implement function to generate codes

### PR DESCRIPTION
**Pull request by wizzy25**
affects: @commercetools/discount-code-generator

#### Summary

Implement the discount code generator function as part of [this](https://github.com/commercetools/nodejs/issues/142#issue-224808585) checklist

### CODE REVIEW

File: packages/discount-code-generator/src/main.js [hunk 1]
```diff
@@ -1,3 +1,62 @@
-export default function findSum (a, b) {
-  return a + b
+/* @flow */
+import TokenGenerator from 'tokgen'
+
+import type {
+  CodeData,
+  CodeDataArray,
+  CodeOptions,
+} from 'types/sdk'
```
Comment (hisabimbola) [inline packages/discount-code-generator/src/main.js]: I think the `types/sdk` file is getting large, separate out your type declarations to another file...

File: packages/discount-code-generator/src/main.js [hunk 2]
```diff
@@ -1,3 +1,62 @@
...
+// The {data} should have the attributes of the discount codes
+// data = {
+//   name: {
+//    en: 'foo'
+//   },
+//   description: {
+//    en: 'bar'
+//   },
+//   cartDiscounts: [],
+//   cartPredicate: 'some predicate',
```
Comment (hisabimbola) [inline packages/discount-code-generator/src/main.js]: Use multi-line comment in javascript

File: packages/discount-code-generator/src/main.js [hunk 3]
```diff
@@ -1,3 +1,62 @@
...
+//   isActive: true,
+//   maxApplicationsPerCustomer: 10,
+//   maxApplicationsPerCustomer: 2,
+//  }
+
+export default function discountCodeGenerator (
+  options: CodeOptions,
```
Comment (hisabimbola) [inline packages/discount-code-generator/src/main.js]: You can wrap all params in one object, except there is possibiliity you reuse each later

File: packages/discount-code-generator/src/main.js [hunk 4]
```diff
@@ -1,3 +1,62 @@
...
+    codes.push(codeObject)
+  }
+  return codes
+}
+
+function _prepareCode (code: string, length: number, prefix: string): string {
+  if (!prefix.length)
+    return code
+
+  const codePrefix = prefix.toUpperCase()
```
Comment (hisabimbola) [inline packages/discount-code-generator/src/main.js]: why you forcing to uppercase?

Comment (hisabimbola) [inline packages/discount-code-generator/src/main.js]: Can the discount codes be in lowercase or mixed?

File: packages/discount-code-generator/test/main.spec.js [hunk 5]
```diff
@@ -1,11 +1,61 @@
...
+  })
+
+  test('should generate codes according to required specifications', () => {
+    const codes = discountCodeGenerator(options, data)
+    expect(Array.isArray(codes)).toBeTruthy()
+    expect(codes.length).toBe(100)
+
+    codes.forEach((codeObject) => {
+      expect(codeObject).toMatchObject(data)
+      expect(codeObject.code).toBeTruthy()
```
Comment (hisabimbola) [inline packages/discount-code-generator/test/main.spec.js]: The next assertion made this assertion redundant

File: packages/discount-code-generator/src/main.js (hunk 3 above)
Comment (wizzy25) [inline packages/discount-code-generator/src/main.js]: Params are currently in one `options` object. The `data` object is the one being mutated

File: packages/discount-code-generator/src/main.js (hunk 4 above)
Comment (wizzy25) [inline packages/discount-code-generator/src/main.js]: Semantics 😇😇

File: packages/discount-code-generator/test/main.spec.js (hunk 5 above)
Comment (wizzy25) [inline packages/discount-code-generator/test/main.spec.js]: The [next assertion](https://github.com/commercetools/nodejs/pull/147/files/e742d5c7fbfa08338c92874e73618914e28ca829#diff-c8f4a8fbd9412cfe0adf807a27b5a09bR54) tests the default values.

**Comment by wizzy25** [comment]
@hisabimbola updated changes, kindly review :)

File: packages/discount-code-generator/src/main.js [hunk 6]
```diff
@@ -1,3 +1,63 @@
...
+ *   name: {
+ *    en: 'foo'
+ *   },
+ *   description: {
+ *    en: 'bar'
+ *   },
+ *   cartDiscounts: [],
+ *   cartPredicate: 'some predicate',
+ *   isActive: true,
+ *   maxApplicationsPerCustomer: 10,
```
Comment (emmenko) [inline packages/discount-code-generator/src/main.js]: This is called `maxApplications`

http://dev.commercetools.com/http-api-projects-discountCodes.html#discountcode

File: types/discountCodes.js [hunk 7]
```diff
@@ -0,0 +1,22 @@
+/* @flow */
+
+/* Discount Codes */
+export type CodeData = {
+  name?: Object;
+  description?: Object;
+  cartDiscounts: [];
+  cartPredicate?: string;
+  isActive: boolean;
+  maxApplicationsPerCustomer?: number;
```
Comment (emmenko) [inline types/discountCodes.js]: This is called `maxApplications`

http://dev.commercetools.com/http-api-projects-discountCodes.html#discountcode

File: packages/discount-code-generator/src/main.js [hunk 8]
```diff
@@ -1,3 +1,63 @@
...
+ * Will throw if no quantity `number` is specified
+ * Defaults to 11-digit codes if no length `number` is specified
+ * Can also specify an optional prefix `string`
+ * options = {
+ *   quantity: 100,
+ *   length: 15,
+ *   prefix: CT,
+ * }
+
+ * The {data} should have the attributes of the discount codes
```
Comment (emmenko) [inline packages/discount-code-generator/src/main.js]: Provide a link to the documentation

http://dev.commercetools.com/http-api-projects-discountCodes.html#discountcode

File: packages/discount-code-generator/src/main.js [hunk 9]
```diff
@@ -1,3 +1,66 @@
...
+ *   },
+ *   cartDiscounts: [],
+ *   cartPredicate: 'some predicate',
+ *   isActive: true,
+ *   maxApplications: 10,
+ *   maxApplicationsPerCustomer: 2,
+ *  }
+ *
+ *  More information about the discount codes can be found here: http://
+ *  dev.commercetools.com/http-api-projects-discountCodes.html#discountcode
```
Comment (wizzy25) [inline packages/discount-code-generator/src/main.js]: Will this be the best way to add this link without annoying eslint? 🤔 Cc @hisabimbola @emmenko

Comment (emmenko) [inline packages/discount-code-generator/src/main.js]: You can simply disable eslint for such comments ;)

**Pull request merged**